std/file.d

```diff
@@ -672,20 +653,202 @@ version(Windows) private void writeImpl(const(char)[] name, const(FSChar)* namez
+                 extension).
+        buffer = Data to populate the file with.
+
+    Returns:
+        The name of the generated file.
+
+    Throws:
+        $(LREF FileException) if it fails to create the file.
+
+    See_Also:
+        $(REF File.tmpfile, std, stdio)
+  +/
+string createTempFile(const void[] buffer = null) @safe
+{
+    return createTempFile(null, null, buffer);
+}
+
+/// Ditto
+string createTempFile(const(char)[] prefix, const(char)[] suffix, const void[] buffer = null) @trusted
+{
+    import std.path : absolutePath, baseName, buildPath, dirName;
+
+    static string genTempName(const(char)[] prefix, const(char)[] suffix)
+    {
+        import std.ascii : digits, letters;
+        import std.random : choice, rndGen;
+        import std.range : chain;
+        import std.string : representation;
+
+        auto name = new char[](prefix.length + 15 + suffix.length);
+        name[0 .. prefix.length] = prefix;
+        name[$ - suffix.length .. $] = suffix;
+
+        auto random = &rndGen();
+        rndGen.popFront();
+
+        auto chars = chain(letters.representation, digits.representation);
+        foreach (ref c; name[prefix.length .. $ - suffix.length])
+        {
+            c = choice(chars);
+            random.popFront();
+        }
+
+        return buildPath(tempDir, name);
+    }
+
+    while (1)
+    {
+        auto filename = genTempName(prefix, suffix);
+
+        version(Posix)
+        {
+            auto fd = open(tempCString!FSChar(filename), O_RDWR | O_CREAT | O_EXCL, S_IRUSR | S_IWUSR);
+            if (fd == -1)
+            {
+                immutable errno = .errno;
+                if (errno == EEXIST)
+                    continue;
+                else
+                    throw new FileException("Failed to create a temporary file", errno);
+            }
+        }
+        else version(Windows)
+        {
+            auto fd = CreateFileW(tempCString!FSChar(filename), GENERIC_WRITE, 0, null, CREATE_NEW,
+                                  FILE_ATTRIBUTE_NORMAL | FILE_FLAG_SEQUENTIAL_SCAN, HANDLE.init);
+            if (fd == INVALID_HANDLE_VALUE)
+            {
+                immutable errno = .GetLastError();
+                if (errno == ERROR_FILE_EXISTS)
+                    continue;
+                else
+                    throw new FileException("Failed to create a temporary file", errno);
+            }
+        }
+        else
+            static assert(0, "Unsupported OS");
+
+        writeToOpenFile(fd, buffer, filename, null);
+
+        return filename;
     }
-    cenforce(sum == size && CloseHandle(h), name, namez);
 }
 
+///
+@safe unittest
+{
+    import std.algorithm.searching : startsWith;
+    import std.range.primitives;
+    import std.path : baseName, buildNormalizedPath, dirName, extension;
+
+    {
+        auto path = createTempFile();
+        scope(exit) if (path.exists) remove(path);
```

Comment:
Why `if (path.exists)`. Should this either be `scope(exit) assert(path.exists), remove(path);` or `scope(exit) remove(path);`?

Reply:
The `path.exists` portion could be removed, but I pretty much always use it before removing a file so that I don't have to worry about `remove` throwing due to something that happened when the code is changed later or something else unforeseen goes wrong.

Reply:
If `tempFile` _does_ fail to create the file for some reason, then one of the tests after the scope statement will fail, and the `AssertError` will bubble up properly, but if the test isn't there, then in that case, we'd then get a `FileException` being thrown instead when `remove` failed. So, IMHO, it is better to have the check even if it's useless if everything is working correctly.